corehq/motech/repeaters/repeater_generators.py

```diff
@@ -261,6 +270,103 @@ def get_test_payload(self, domain):
         )
 
 
+@attr.s
+class CaseTypeReferralConfig(object):
+    #if false, listed_properties is a whitelist
+    use_blacklist = attr.ib()
+    listed_properties = attr.ib()
+    constant_properties = attr.ib()
+
+
+class ReferCasePayloadGenerator(BasePayloadGenerator):
+
+    def get_payload(self, repeat_record, payload_doc):
+
+        def _update_case_id(original_case_id, case_id_map):
+            if original_case_id in case_id_map:
+                new_case_id = case_id_map[original_case_id]
+            else:
+                new_case_id = uuid4().hex
+                case_id_map[original_case_id] = new_case_id
+            return new_case_id
+
+        def _update_case_properties_with_blacklist(case, config):
+            for name in config.listed_properties:
+                if name in case.case_json:
+                    del case.case_json[name]
+
+        def _update_case_properties_with_whitelist(case, config):
+            new_json = {}
+            for name in config.listed_properties:
+                if name in case.case_json:
+                    new_json[name] = case.case_json[name]
+            case.case_json = new_json
+
+        def _set_constant_properties(case, config):
+            for name, value in config.constant_properties:
+                case.case_json[name] = value
+
+        case_ids_to_forward = payload_doc.get_case_property('cases_to_forward').split(' ')
```

Comment:
This section, until the blank line in 339, sets up the configuration for copying the cases, including all the properties that need to be copied, added, or changed